### PR TITLE
Add OpenTracing configurability for APIcast environments

### PR DESCRIPTION
**Pull request by miguelsorianod**
This exposes OpenTracing functionality provided by APIcast into the APIManager CR for the APIcast staging and production environments.

### Design

New `spec.apicast.stagingSpec.openTracing` and `spec.apicast.produtionSpec.openTracing` sections have been added in the `APIManager` CR.

An example:

```
spec:
  apicast:
    stagingSpec:
      ...
      openTracing:
        enabled: true
        tracingLibrary: jaeger
        tracingConfigRef:
          name: mysecretname
    productionSpec:
       ...
      openTracing:
        enabled: true
        tracingLibrary: jaeger
        tracingConfigRef:
          name: mysecretname
  ...
```

* All attributes in the `openTracing` section are optional
* If the `enabled` flag is not specified, it defaults to `false`
* If the `tracingLibrary` flag is not specified, it defaults to `jaeger`
* If the `tracingConfigRef` flag is not specified, it defaults to a default tracing configuration specific to the set `tracingLirary`. This default tracing configuration is provided in the APIcast image itself

The configuration is mounted into the APIcast pod by looking at a `config` key in the provided secret name and mounted in `/opt/app-root/src/tracing-configs/tracing-config-<tracing_library_name>-<secret_name>`, for example in `opt/app-root/src/tracing-configs/tracing-config-jaeger-mysecretname`

The volume tracking annotation mechanism has been implemented for the opentracing-related Volumes in APIcast DeploymentConfigs in order to reconcile addition/removal of them without altering manually mounted ones.

The annotation format for them is:

```
apps.3scale.net/tracing-config-volume-<md5sum-of-volume-name>: <volume-name>
```

The Volume name format is:
```
tracing-config-<tracing_library>-<tracing_config_secret_name>
```

### CODE REVIEW

File: pkg/3scale/amp/component/apicast.go [hunk 1]
```diff
@@ -350,6 +358,18 @@ func (apicast *Apicast) buildApicastStagingEnv() []v1.EnvVar {
...
 	}
+
+	stagingTracingConfig := apicast.Options.StagingTracingConfig
+	if stagingTracingConfig.Enabled {
+		result = append(result, helper.EnvVarFromValue("OPENTRACING_TRACER", stagingTracingConfig.TracingLibrary))
+
+		if stagingTracingConfig.TracingConfigSecretName != nil {
+			result = append(result,
+				helper.EnvVarFromValue("OPENTRACING_CONFIG",
+					strings.Join([]string{APIcastTracingConfigMountBasePath, stagingTracingConfig.VolumeName()}, "/")))
```
Comment (eguzki) [inline pkg/3scale/amp/component/apicast.go:369]: `path.Join` ?

File: pkg/3scale/amp/component/apicast.go [hunk 2]
```diff
@@ -367,6 +387,20 @@ func (apicast *Apicast) buildApicastProductionEnv() []v1.EnvVar {
...
 	}
+
+	productionTracingConfig := apicast.Options.ProductionTracingConfig
+	if productionTracingConfig.Enabled {
+		result = append(result, helper.EnvVarFromValue("OPENTRACING_TRACER", productionTracingConfig.TracingLibrary))
+
+		if productionTracingConfig.TracingConfigSecretName != nil {
+			result = append(result,
+				helper.EnvVarFromValue("OPENTRACING_CONFIG",
+					strings.Join([]string{APIcastTracingConfigMountBasePath,
```
Comment (eguzki) [inline pkg/3scale/amp/component/apicast.go:398]: `path.Join` ?

File: pkg/3scale/amp/component/apicast_options.go [hunk 3]
```diff
@@ -26,13 +26,38 @@ func (c CustomPolicy) AnnotationKey() string {
 	// prefix/name: value
 	// The name segment is required and must be 63 characters or less
 	// Currently: len(CustomPoliciesAnnotationKeyPrefix) + 32 (from the hash) = 54
```
Comment (eguzki) [inline pkg/3scale/amp/component/apicast_options.go:28]: change the comment as well

File: pkg/3scale/amp/component/apicast_options.go [hunk 4]
```diff
@@ -26,13 +26,38 @@ func (c CustomPolicy) AnnotationKey() string {
...
+	return fmt.Sprintf("tracing-config-%s-%s", helper.DNS1123Name(c.TracingLibrary), *c.TracingConfigSecretName)
+}
+
+// AnnotationKey returns the annotation key associated to the tracing config volume name. It should only be used
+// when c.TracingConfigSecretName is not nil
+func (c APIcastTracingConfig) AnnotationKey() string {
+	// https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/
+	// prefix/name: value
+	// The name segment is required and must be 63 characters or less
+	// Currently: len(APIcastTracingConfigAnnotationNameSegmentPrefix) + 32 (from the hash) + "-" = 62
```
Comment (eguzki) [inline pkg/3scale/amp/component/apicast_options.go:57]: living in the edge

File: pkg/3scale/amp/operator/apicast_options_provider.go [hunk 5]
```diff
@@ -202,3 +209,119 @@ func (a *ApicastOptionsProvider) validateCustomPolicySecret(name string) error {
...
+func (a *ApicastOptionsProvider) setProductionTracingConfiguration() error {
+	tracingIsEnabled := a.apimanager.IsAPIcastProductionOpenTracingEnabled()
+	res := &component.APIcastTracingConfig{
+		Enabled:        tracingIsEnabled,
+		TracingLibrary: component.APIcastDefaultTracingLibrary,
+	}
+	if tracingIsEnabled {
+		openTracingConfigSpec := a.apimanager.Spec.Apicast.ProductionSpec.OpenTracing
+		if openTracingConfigSpec.TracingLibrary != nil {
+			// For now only "jaeger" is accepted" as the tracing library
```
Comment (eguzki) [inline pkg/3scale/amp/operator/apicast_options_provider.go:236]: This check seems to belong to APIManager CR `Validate()` method. IT is being validated before trying reconciliation. The option provider is executed as part of the reconciliation.

**Review by eguzki — CHANGES_REQUESTED** [review]
Overall looking good. I dropped some comments

File: pkg/3scale/amp/component/apicast.go (hunk 1 above)
Comment (miguelsorianod) [reply pkg/3scale/amp/component/apicast.go]: changed

File: pkg/3scale/amp/component/apicast.go (hunk 2 above)
Comment (miguelsorianod) [reply pkg/3scale/amp/component/apicast.go]: changed

File: pkg/3scale/amp/component/apicast_options.go (hunk 3 above)
Comment (miguelsorianod) [reply pkg/3scale/amp/component/apicast_options.go:28]: Unfortunately with the current code for custom environment there's no variable that reflects that part unless we split it as I've done with the opentracing annotation. That comment was originally wrong already.
If you want I can split it as I've done with the opentracing config to make it accurate

File: pkg/3scale/amp/operator/apicast_options_provider.go (hunk 5 above)
Comment (miguelsorianod) [reply pkg/3scale/amp/operator/apicast_options_provider.go]: changed :+1:

File: pkg/3scale/amp/component/apicast_options.go (hunk 3 above)
Comment (eguzki) [reply pkg/3scale/amp/component/apicast_options.go:28]: Ok, I will fix that in the other PR about custom envs. 

Yes, I am aware that originally the comment was wrong.

**Review by eguzki — APPROVED** [review]
Some doc about how to see opentracing result when enabled?

**Comment by miguelsorianod** [comment]
> Some doc about how to see opentracing result when enabled?

That depends on how they install opentracing. They should see the traces in the opentracing ui if successfully working

File: pkg/3scale/amp/component/apicast_options.go (hunk 3 above)
Comment (miguelsorianod) [reply pkg/3scale/amp/component/apicast_options.go]: :+1: updated

**Pull request merged**